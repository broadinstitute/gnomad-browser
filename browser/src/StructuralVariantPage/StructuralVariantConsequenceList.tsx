import React, { Component } from 'react'
import styled from 'styled-components'

import { ListItem, Modal, OrderedList, TextButton } from '@gnomad/ui'

import InfoButton from '../help/InfoButton'
import Link from '../Link'
import {
  svConsequenceCategories,
  svConsequenceLabels,
} from '../StructuralVariantList/structuralVariantConsequences'
import { StructuralVariant } from './StructuralVariantPage'

const Wrapper = styled.ol`
  display: flex;
  flex-flow: row wrap;
  padding: 0;
  list-style-type: none;
  margin-bottom: 1em;

  h3,
  h4 {
    margin: 0 0 0.5em;
  }
`

const ConsequenceListItem = styled.li`
  flex-basis: 220px;
`

type Props = {
  variant: StructuralVariant
}

type State = any

class StructuralVariantConsequenceList extends Component<Props, State> {
  state = {
    expandedConsequence: null,
  }

  render() {
    const { variant } = this.props
    const { expandedConsequence } = this.state
    const consequences = variant.consequences || []
    return (
      <React.Fragment>
        <Wrapper>
          {consequences.map((consequence) => {
            const consequenceCode = consequence.consequence
            const genes = consequence.genes || []

            const category = svConsequenceCategories[consequenceCode]

            // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
            const helpTopic = {
              lof: 'sv-effect_pLoF',
              dup_lof: 'sv-effect_IED',
              copy_gain: 'sv-effect_CG',
            }[category]

            return (
              <ConsequenceListItem key={consequenceCode}>
                <h3>
                  {svConsequenceLabels[consequenceCode]}{' '}
                  {!!helpTopic && <InfoButton topic={helpTopic} />}
                </h3>
                {/* @ts-expect-error TS(2746) FIXME: This JSX tag's 'children' prop expects a single ch... Remove this comment to see the full error message */}
                <OrderedList>
                  {genes.slice(0, 3).map((gene) => (
                    // @ts-expect-error TS(2769) FIXME: No overload matches this call.
                    <ListItem key={gene}>
                      <Link to={`/gene/${gene}`}>{gene}</Link>
                    </ListItem>
                  ))}
                  {genes.length > 3 && (
                    // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
                    <ListItem>
                      <TextButton
                        onClick={() => {
                          this.setState({ expandedConsequence: category })
                        }}
                      >
                        and {genes.length - 3} more
                      </TextButton>
                    </ListItem>
                  )}
                </OrderedList>
              </ConsequenceListItem>
            )
          })}
        </Wrapper>
        {expandedConsequence && (
          // @ts-expect-error TS(2741) FIXME: Property 'size' is missing in type '{ children: El... Remove this comment to see the full error message
          <Modal
            title={`${svConsequenceLabels[expandedConsequence]}`}
            onRequestClose={() => {
              this.setState({ expandedConsequence: null })
            }}
          >
            {/* @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
            <OrderedList>
              {/* @ts-expect-error TS(2532) FIXME: Object is possibly 'undefined'. */}
              {variant.consequences
                .find(({ consequence }: any) => consequence === expandedConsequence)
                .genes.map((gene: any) => (
                  // @ts-expect-error TS(2769) FIXME: No overload matches this call.
                  <ListItem key={gene}>
                    <Link to={`/gene/${gene}`}>{gene}</Link>
                  </ListItem>
                ))}
            </OrderedList>
          </Modal>
        )}
      </React.Fragment>
    )
  }
}

export default StructuralVariantConsequenceList
