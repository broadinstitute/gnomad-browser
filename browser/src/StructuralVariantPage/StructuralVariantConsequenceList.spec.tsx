import React from 'react'
import { render } from '@testing-library/react'
import { expect, test, describe } from '@jest/globals'
import svFactory from '../__factories__/StructuralVariant'
import StructuralVariantConsequenceList from './StructuralVariantConsequenceList'
import { BrowserRouter } from 'react-router-dom'

describe('StructuralVariantConsequenceList', () => {
  test('should have no unexpected changes', () => {
    const consequences = [
      { consequence: 'lof', genes: null },
      { consequence: 'lof', genes: ['ABC123'] },
      { consequence: 'lof', genes: ['ABC123', 'QRSTUV'] },
      { consequence: 'copy_gain', genes: null },
      { consequence: 'copy_gain', genes: ['DEF234'] },
      { consequence: 'copy_gain', genes: ['DEF234', 'LLLLLL'] },
    ]
    const variant = svFactory.build({ consequences })
    const { asFragment } = render(
      <BrowserRouter>
        <StructuralVariantConsequenceList variant={variant} />
      </BrowserRouter>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
