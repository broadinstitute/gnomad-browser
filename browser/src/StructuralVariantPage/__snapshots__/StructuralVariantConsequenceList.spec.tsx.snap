// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructuralVariantConsequenceList expanding genes list for consequence in category copy_gain with >3 genes expands to show all genes 1`] = `
<DocumentFragment>
  <ol
    class="StructuralVariantConsequenceList__Wrapper-sc-1mstxep-0 fKPQdv"
  >
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        copy gain 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/ABC"
          >
            ABC
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/DEF"
          >
            DEF
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/GHI"
          >
            GHI
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <button
            class="Button__TextButton-sc-1eobygi-3 gWkUTz"
            type="button"
          >
            and 1 more
          </button>
        </li>
      </ol>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`StructuralVariantConsequenceList expanding genes list for consequence in category dup_lof with >3 genes expands to show all genes 1`] = `
<DocumentFragment>
  <ol
    class="StructuralVariantConsequenceList__Wrapper-sc-1mstxep-0 fKPQdv"
  >
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        int. exon duplication 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/ABC"
          >
            ABC
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/DEF"
          >
            DEF
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/GHI"
          >
            GHI
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <button
            class="Button__TextButton-sc-1eobygi-3 gWkUTz"
            type="button"
          >
            and 1 more
          </button>
        </li>
      </ol>
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        partial exon duplication 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/MNO"
          >
            MNO
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/PQR"
          >
            PQR
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/STU"
          >
            STU
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <button
            class="Button__TextButton-sc-1eobygi-3 gWkUTz"
            type="button"
          >
            and 2 more
          </button>
        </li>
      </ol>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`StructuralVariantConsequenceList expanding genes list for consequence in category lof with >3 genes expands to show all genes 1`] = `
<DocumentFragment>
  <ol
    class="StructuralVariantConsequenceList__Wrapper-sc-1mstxep-0 fKPQdv"
  >
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        loss of function 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/ABC"
          >
            ABC
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/DEF"
          >
            DEF
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/GHI"
          >
            GHI
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <button
            class="Button__TextButton-sc-1eobygi-3 gWkUTz"
            type="button"
          >
            and 1 more
          </button>
        </li>
      </ol>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`StructuralVariantConsequenceList expanding genes list for consequence in category other with >3 genes expands to show all genes 1`] = `
<DocumentFragment>
  <ol
    class="StructuralVariantConsequenceList__Wrapper-sc-1mstxep-0 fKPQdv"
  >
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        inversion span 
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/ABC"
          >
            ABC
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/DEF"
          >
            DEF
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/GHI"
          >
            GHI
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <button
            class="Button__TextButton-sc-1eobygi-3 gWkUTz"
            type="button"
          >
            and 1 more
          </button>
        </li>
      </ol>
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        partial duplication 
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/MNO"
          >
            MNO
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/PQR"
          >
            PQR
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/STU"
          >
            STU
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <button
            class="Button__TextButton-sc-1eobygi-3 gWkUTz"
            type="button"
          >
            and 2 more
          </button>
        </li>
      </ol>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`StructuralVariantConsequenceList should have no unexpected changes 1`] = `
<DocumentFragment>
  <ol
    class="StructuralVariantConsequenceList__Wrapper-sc-1mstxep-0 fKPQdv"
  >
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        loss of function 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      />
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        loss of function 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/ABC123"
          >
            ABC123
          </a>
        </li>
      </ol>
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        loss of function 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/ABC123"
          >
            ABC123
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/QRSTUV"
          >
            QRSTUV
          </a>
        </li>
      </ol>
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        copy gain 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      />
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        copy gain 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/DEF234"
          >
            DEF234
          </a>
        </li>
      </ol>
    </li>
    <li
      class="StructuralVariantConsequenceList__ConsequenceListItem-sc-1mstxep-1 qVCRP"
    >
      <h3>
        copy gain 
        <button
          class="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            More information
          </span>
        </button>
      </h3>
      <ol
        class="List__OrderedList-sc-4j87st-2 gKjCRF"
      >
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/DEF234"
          >
            DEF234
          </a>
        </li>
        <li
          class="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            class="Link-sc-14lgydv-0-Link csqFBR"
            href="/gene/LLLLLL"
          >
            LLLLLL
          </a>
        </li>
      </ol>
    </li>
  </ol>
</DocumentFragment>
`;
